rkt/fetch_keys.go

```diff
@@ -0,0 +1,237 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/appc/spec/discovery"
+	"github.com/coreos/rkt/Godeps/_workspace/src/golang.org/x/crypto/openpgp"
+)
+
+// getPubKeyLocations either returns the locations supplied in argv or discovers one at prefix
+func getPubKeyLocations(prefix string, allowHTTP bool, debug bool) ([]string, error) {
+	if prefix == "" {
+		return nil, fmt.Errorf("at least one key or --prefix required")
+	}
+
+	kls, err := metaDiscoverPubKeyLocations(prefix, allowHTTP, debug)
+	if err != nil {
+		return nil, fmt.Errorf("--prefix meta discovery error: %v", err)
+	}
+
+	if len(kls) == 0 {
+		return nil, fmt.Errorf("meta discovery on %s resulted in no keys", prefix)
+	}
+
+	return kls, nil
+}
+
+// metaDiscoverPubKeyLocations discovers the public key through ACDiscovery by applying prefix as an ACApp
+func metaDiscoverPubKeyLocations(prefix string, allowHTTP bool, debug bool) ([]string, error) {
+	app, err := discovery.NewAppFromString(prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	ep, attempts, err := discovery.DiscoverPublicKeys(*app, allowHTTP)
+	if err != nil {
+		return nil, err
+	}
+
+	if debug {
+		for _, a := range attempts {
+			stderr("meta tag 'ac-discovery-pubkeys' not found on %s: %v", a.Prefix, a.Error)
+		}
+	}
+
+	return ep.Keys, nil
+}
+
+// getPubKey retrieves a public key (if remote), and verifies it's a gpg key
+func getPubKey(location string, allowHTTP bool) (*os.File, error) {
+	u, err := url.Parse(location)
+	if err != nil {
+		return nil, err
+	}
+
+	switch u.Scheme {
+	case "":
+		return os.Open(location)
+	case "http":
+		if !allowHTTP {
+			return nil, fmt.Errorf("--insecure-allow-http required for http URLs")
+		}
+		fallthrough
+	case "https":
+		return downloadKey(u.String())
+	}
+
+	return nil, fmt.Errorf("only http and https urls supported")
+}
+
+// downloadKey retrieves the file, storing it in a deleted tempfile
+func downloadKey(url string) (*os.File, error) {
+	tf, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("error creating tempfile: %v", err)
+	}
+	os.Remove(tf.Name()) // no need to keep the tempfile around
+
+	defer func() {
+		if err != nil {
+			tf.Close()
+		}
+	}()
+
+	res, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("error getting key: %v", err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("bad HTTP status code: %d", res.StatusCode)
+	}
+
+	if _, err := io.Copy(tf, res.Body); err != nil {
+		return nil, fmt.Errorf("error copying key: %v", err)
+	}
+
+	tf.Seek(0, os.SEEK_SET)
+
+	return tf, nil
+}
+
+// addKeys adds the keys listed in pkls at prefix
+func addKeys(pkls []string, prefix string, allowHTTP bool, forceAccept bool, allowOverride bool) error {
```

Comment:
You can shorten this line by omitting `bool` after `allowHTTP` and `forceAccept` parameters.

Also, I really don't like `bool` parameters. Seeing an invocation like `addKeys(foo, "bar", true, false, true)` requires you to actually lookup a prototype of the function to see what all those bool are about.